src/nunit.analyzers/DisposeFieldsInTearDown/DisposeFieldsInTearDownAnalyzer.cs

```diff
@@ -0,0 +1,442 @@
+
+            foreach (var method in methods)
+            {
+                HashSet<string> assignedSymbolsInMethod = AssignedIn(model, type, symbols, method);
+                assignedSymbols.UnionWith(assignedSymbolsInMethod);
+            }
+
+            return assignedSymbols;
+        }
+
+        /// <summary>
+        /// Returns a hash set of the fields assigned in <paramref name="symbol"/>.
+        /// </summary>
+        /// <param name="symbol">The method to look for.</param>
+        /// <param name="symbols">The symbols to check for assignment.</param>
+        /// <returns>HashSet of <paramref name="symbols"/> that are assigned in <paramref name="symbol"/>.</returns>
+        private static HashSet<string> AssignedIn(SemanticModel model, INamedTypeSymbol type, HashSet<string> symbols, IMethodSymbol symbol)
+        {
+            MethodDeclarationSyntax? method =
+                symbol.DeclaringSyntaxReferences.FirstOrDefault()?.GetSyntax() as MethodDeclarationSyntax;
+
+            return method is null ? new HashSet<string>() : AssignedIn(model, type, symbols, method);
+        }
+
+        private static HashSet<string> AssignedIn(SemanticModel model, INamedTypeSymbol type, HashSet<string> symbols, MethodDeclarationSyntax method)
+        {
+            if (method.ExpressionBody is not null)
+            {
+                return AssignedIn(model, type, symbols, method.ExpressionBody.Expression);
+            }
+
+            if (method.Body is not null)
+            {
+                return AssignedIn(model, type, symbols, method.Body);
+            }
+
+            return new HashSet<string>();
+        }
+
+        private static HashSet<string> AssignedIn(SemanticModel model, INamedTypeSymbol type, HashSet<string> symbols, ExpressionSyntax expression)
+        {
+            var assignedSymbols = new HashSet<string>();
+            if (expression is AssignmentExpressionSyntax assignmentExpression)
+            {
+                // We only deal with simple assignments, not tuple or deconstruct
+                string? name = GetIdentifier(assignmentExpression.Left);
+                if (name is not null && symbols.Contains(name))
+                {
+                    TypeInfo typeInfo = model.GetTypeInfo(assignmentExpression.Right);
+                    if (typeInfo.Type?.IsDisposable() == true)
+                    {
+                        // Make one exemption, if the value is returned from a 'xxx.Add()' call.
+                        // It is then assumed that owner ship is transferred to that 'collection'.
+                        // This matches the (undocumented) CA2000 behaviour.
+                        if (assignmentExpression.Right is not InvocationExpressionSyntax invocationExpression ||
+                            invocationExpression.Expression is not MemberAccessExpressionSyntax memberAccessExpression ||
+                            memberAccessExpression.Name.Identifier.Text != "Add")
```

Comment:
The only thing official is the source code: [`IsCollectionAddMethod`](https://github.com/dotnet/roslyn-analyzers/blob/main/src/Utilities/Compiler/Extensions/IMethodSymbolExtensions.cs#L465-L499). From that, the method should be either `Add` or `AddOrUpdate` even though the latter is not part of `ICollection`. Go figure.

Added the above link in the comment.

Reply:
As far as I can tell from the code you linked to it can either be one of the well-known names or a name starting with `.Add`, but I don't think this is very important (given that I didn't know anything about this before this PR :) )

```
        /// We use the following heuristic to determine if a method is a collection add method:
        /// 1. Method's enclosing type implements any of the given <paramref name="iCollectionTypes"/>.
        /// 2. Any of the following name heuristics are met:
        ///     a. Method's name is from one of the well-known add method names from <see cref="s_collectionAddMethodNameVariants"/> ("Add", "AddOrUpdate", "GetOrAdd", "TryAdd", or "TryUpdate")
        ///     b. Method's name begins with "Add" (FxCop compat)
```

Reply:
Thanks, you are correct, I read the condition wrong. I saw `&&` but missed the `!`.